### DISCUSSION TITLE
Source information do not appear

### DISCUSSION THREAD

**Discussion by chengm349** [gabime/spdlog#2239] [Q&A]
I know there is a link https://spdlog.docsforge.com/v1.x/0.faq/#source-information-do-not-appear-when-using-custom-format talking about this restriction.

#define SPDLOG_INFO_VERBOSE(...) SPDLOG_LOGGER_CALL(spdlog::default_logger_raw(), spdlog::level::info, __VA_ARGS__)
#define SPDLOG_TRACE_VERBOSE(...) SPDLOG_LOGGER_CALL(spdlog::default_logger_raw(), spdlog::level::trace, __VA_ARGS__)

spdlog::set_level(spdlog::level::trace);
SPDLOG_TRACE_VERBOSE("Welcome to spdlog!");
spdlog::info("Welcome to spdlog!");
auto console = spd::stdout_color_mt("console");
spdlog::set_pattern("[%H:%M:%S %f] [%t] [%^%l%$] %v");
SPDLOG_INFO_VERBOSE("Welcome to spdlog!");
console->error("Some error message with arg{}..", 1);

output:
[2022-01-07 18:15:28.120] [trace] [spdlogTest.cpp:42] Welcome to spdlog!
[2022-01-07 18:15:28.120] [info] Welcome to spdlog!
[18:15:28 120806] [18013] [info] Welcome to spdlog!
[18:15:28 120906] [18013] [error] Some error message with arg1..
[18:15:28 120949] [18013] [warning] Easy padding in numbers like 00000012

Keen to know why I can't call set_pattern().
I want source information and dynamic level setting by spdlog::set_level()

**Comment by tt4g** [comment]
That document is wrong. See FAQ: https://github.com/gabime/spdlog/wiki/0.-FAQ#source-information-do-not-appear-when-using-custom-format
Must be define `SPDLOG_ACTIVE_LEVEL` macro.

Example:

```cpp
#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_INFO 
#include "spdlog/spdlog.h"
#include "spdlog/sinks/stdout_sinks.h"

void source_info_example()
{
    auto console = spdlog::stdout_logger_mt("console");
    spdlog::set_default_logger(console);
    spdlog::set_pattern("[source %s] [function %!] [line %#] %v");

    SPDLOG_LOGGER_INFO(console, "log with source info"); // Console: "[source example.cpp] [function source_info_example] [line 10] log with source info"
    SPDLOG_INFO("global log with source info"); // Console: "[source example.cpp] [function source_info_example] [line 11] global logger with source info"

    console->info("source info is not printed"); // Console: "[source ] [function ] [line ] source info is not printed"
}
```

**Comment by chengm349** [comment]
If I add these lines:

#define SPDLOG_LEVEL_NAMES \
    { spdlog::string_view_t("trace", 5)     \
    , spdlog::string_view_t("debug", 5)     \
    , spdlog::string_view_t("info", 4)      \
    , spdlog::string_view_t("warn", 4)   \
    , spdlog::string_view_t("error", 5)     \
    , spdlog::string_view_t("fatal", 5)  \
    , spdlog::string_view_t("off", 3) }
#define SPDLOG_ACTIVE_LEVEL  **SPDLOG_LEVEL_TRACE**
#include "spdlog/spdlog.h"
[result.txt](https://github.com/gabime/spdlog/files/7832514/result.txt)


I got some compilation error (result.txt).

If I change to #define SPDLOG_ACTIVE_LEVEL  **SPDLOG_LEVEL_INFO**

Then I can compile ok.

**Reply by tt4g** [reply]
The only thing I could see from the compile error was that you were using the `SPDLOG_***` macro wrong.
Please provide us with source code that can reproduce the problem.

The following code works for me:

```cpp
#define SPDLOG_LEVEL_NAMES \
    { spdlog::string_view_t("trace", 5) \
    , spdlog::string_view_t("debug", 5) \
    , spdlog::string_view_t("info", 4) \
    , spdlog::string_view_t("warning", 7) \
    , spdlog::string_view_t("error", 5) \
    , spdlog::string_view_t("fatal", 5) \
    , spdlog::string_view_t("off", 3) }

#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_TRACE

#include "spdlog/spdlog.h"
#include "spdlog/sinks/stdout_sinks.h"

int main(void)
{
    spdlog::set_level(spdlog::level::level_enum::trace);

    auto console = spdlog::stdout_logger_mt("console");
    spdlog::set_default_logger(console);
    spdlog::set_pattern("[source %s] [function %!] [line %#] %v");

    SPDLOG_TRACE("{} log.", "trace");
    SPDLOG_LOGGER_INFO(console, "info with 2 arguments. {} {}", "foo", 100);
    SPDLOG_CRITICAL("critical (alias)");

    return 0;
}
```

**Reply by chengm349** [reply]
yours works. 

I started to use this beautiful lib several years ago with my wrapper (which supports MT) to support stream usage pattern and by that I can control when to add source info; or some users may prefer stream usage pattern.  But recently I noticed that my wrapper also increased the code size. And I improved my wrapper these couple of days and it does help to reduce the size but still bigger than by using the native macro/API.

**Comment by chengm349** [comment]
when I use SPDLOG_INFO, I'd like to use one pattern and while I use spdlog::info, I'd like to use another pattern.  I have not found a nice way to do achieve it.

More meaningful to me is that I want #define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_TRACE to work. When in TRACE, ERROR/Critical level, I want source information.

**Reply by tt4g** [reply]
spdlog does not support the ability to use a different pattern for each log level.
If you need it, please create your own macros.

**Reply by chengm349** [reply]
yes, I'll try. thanks for your quick response.

**Comment by chengm349** [comment]
btw, my problem code
[spdlogTest.zip](https://github.com/gabime/spdlog/files/7832804/spdlogTest.zip)

I think you can easily remove  them:

#include "DepthSource.h"
#include "FtCmnDefs.h"

**Reply by tt4g** [reply]
Fix lines 109-110.

```diff
-        SPDLOG_TRACE(console, "Enabled only #ifdef SPDLOG_TRACE_ON..{} ,{}", 1, 3.23);
-        SPDLOG_DEBUG(console, "Enabled only #ifdef SPDLOG_DEBUG_ON.. {} ,{}", 1, 3.23)
+       SPDLOG_LOGGER_TRACE(console, "Enabled only #ifdef SPDLOG_TRACE_ON..{} ,{}", 1, 3.23);
+       SPDLOG_LOGGER_DEBUG(console, "Enabled only #ifdef SPDLOG_DEBUG_ON.. {} ,{}", 1, 3.23);
```

If the logger is passed as the first argument, you have to use the `SPDLOG_LOGGER_<LEVEL_NAME>` macro.
`SPDLOG_<LEVEL_NAME>` macro cannot pass the logger because it uses `spdlog::default_logger()`.

**Reply by chengm349** [reply]
my bad, wanted to quickly try this, try that but forgot the hold basic things. thanks.
